### PR TITLE
Add support for VT100 Auto Wrap Mode (DECAWM)

### LINKED ISSUE #3826: Add support for VT100 Auto Wrap Mode (DECAWM)
# Description of the new feature/enhancement

The [`DECAWM`](https://vt100.net/docs/vt510-rm/DECAWM.html) escape sequence is a DEC private mode which controls whether characters wrap at the end of the line. It corresponds, more or less, with the existing console `ENABLE_WRAP_AT_EOL_OUTPUT` flag.

I'm not sure how widely it's used, but it's supported by most terminal emulators I've tested, and it's required to pass the _Test of screen features_ in Vttest.

# Proposed technical implementation details

Since we already have the `ENABLE_WRAP_AT_EOL_OUTPUT` flag, it should just be a matter of hooking up the mode change escape sequence to toggle that flag. However, there is one slight catch - disabling the `ENABLE_WRAP_AT_EOL_OUTPUT` flag doesn't actually work when `WC_DELAY_EOL_WRAP` is set (which is always the case in VT mode), so that'll need to be fixed first.

### PR DESCRIPTION
**Pull request by j4james**
## Summary of the Pull Request

This adds support for the [`DECAWM`](https://vt100.net/docs/vt510-rm/DECAWM) private mode escape sequence, which controls whether or not the output wraps to the next line when the cursor reaches the right edge of the screen. Tested manually, with [Vttest](https://invisible-island.net/vttest/), and with some new unit tests.

## PR Checklist
* [x] Closes #3826
* [x] CLA signed. If not, go over [here](https://cla.opensource.microsoft.com/microsoft/Terminal) and sign the CLA
* [x] Tests added/passed
* [ ] Requires documentation to be updated
* [x] I've discussed this with core contributors already. If not checked, I'm ready to accept this work might be rejected in favor of a different grand plan. Issue number where discussion took place: #3826

## Detailed Description of the Pull Request / Additional comments

The idea was to repurpose the existing `ENABLE_WRAP_AT_EOL_OUTPUT` mode, but the problem with that was it didn't work in VT mode - specifically, disabling it didn't prevent the wrapping from happening. This was because in VT mode the `WC_DELAY_EOL_WRAP` behaviour takes affect, and that bypasses the usual codepath where `ENABLE_WRAP_AT_EOL_OUTPUT` is checked,

To fix this, I had to add additional checks in the `WriteCharsLegacy` function (7dbefe06e41f191a0e83cfefe4896b66094c4089) to make sure the `WC_DELAY_EOL_WRAP` mode is only activated when `ENABLE_WRAP_AT_EOL_OUTPUT`  is also set.

Once that was fixed, though, another issue came to light: the `ENABLE_WRAP_AT_EOL_OUTPUT` mode doesn't actually work as documented. According to the docs, "if this mode is disabled, the last character in the row is overwritten with any subsequent characters". What actually happens is the cursor jumps back to the position at the start of the write, which could be anywhere on the line.

This seems completely broken to me, but I've checked in the Windows XP, and it has the same behaviour, so it looks like that's the way it has always been. So I've added a fix for this (9df98497ca38f7d0ea42623b723a8e2ecf9a4ab9), but it is only applied in VT mode.

Once that basic functionality was in place, though, we just needed a private API in the `ConGetSet` interface to toggle the mode, and then that API could be called from the `AdaptDispatch` class when the `DECAWM` escape sequence was received.

One last thing was to reenable the mode in reponse to a `DECSTR` soft reset. Technically the auto wrap mode was disabled by default on many of the DEC terminals, and some documentation suggests that `DECSTR` should reset it to that state, But most modern terminals (including XTerm) expect the wrapping to be enabled by default, and `DECSTR` reenables that state, so that's the behaviour I've copied.

## Validation Steps Performed

I've add a state machine test to confirm the `DECAWM` escape is dispatched correctly, and a screen buffer test to make sure the output is wrapped or clamped as appropriate for the two states.

I've also confirmed that the "wrap around" test is now working correctly in the _Test of screen features_ in Vttest.

### CODE REVIEW

File: src/host/getset.cpp [hunk 1]
```diff
@@ -1291,6 +1291,25 @@ void ApiRoutines::GetConsoleDisplayModeImpl(ULONG& flags) noexcept
...
+//     when it reaches the end of the current row.
+// Parameters:
+// - wrapAtEOL - set to true to wrap, false to overwrite the last character.
+// Return value:
+// - STATUS_SUCCESS if handled successfully.
+[[nodiscard]] NTSTATUS DoSrvPrivateSetAutoWrapMode(const bool wrapAtEOL)
+{
+    auto& gci = ServiceLocator::LocateGlobals().getConsoleInformation();
+    auto& outputMode = gci.GetActiveOutputBuffer().GetActiveBuffer().OutputMode;
+    if (wrapAtEOL)
```
Comment (miniksa) [inline src/host/getset.cpp]: Please use braces, even for one line if/else.

Comment (miniksa) [reply src/host/getset.cpp]: Alternatively, you can use `WI_UpdateFlag` here.

Comment (j4james) [reply src/host/getset.cpp]: Yeah, that's much better. Thanks.

**Comment by j4james** [comment]
> I want to get this in before I work on the boogaloo stream-writing fixes (#780) that are intended to replace WriteCharsLegacy.

If #780 is likely to happen soonish, I'm actually wondering now if we're better off leaving this PR for the time being, and then later reimplement it on top of the new stream writer? At that point, none of this code would be needed in `WriteCharsLegacy` and `AdjustCursorPosition`, so we'd just be pulling it all out again. I don't mind either way.

**Comment by miniksa** [comment]
> > I want to get this in before I work on the boogaloo stream-writing fixes (#780) that are intended to replace WriteCharsLegacy.
> 
> If #780 is likely to happen soonish, I'm actually wondering now if we're better off leaving this PR for the time being, and then later reimplement it on top of the new stream writer? At that point, none of this code would be needed in `WriteCharsLegacy` and `AdjustCursorPosition`, so we'd just be pulling it all out again. I don't mind either way.

Let's bring it in because it will guide the way I work. And I think these in conjunction with your C0 proposal will make #780 more self-evident.

**Comment by j4james** [comment]
(╯‵□′)╯︵┻━┻

**Comment by miniksa** [comment]
> (╯‵□′)╯︵┻━┻

Going to try to re-run/merge until success.

**Comment by miniksa** [comment]
OK after I busted #4171 by clicking update branch, I don't really want to do it here. It looks like the build cannot pull the merge result branch. Can you update your branch, @j4james?

**Comment by miniksa** [comment]
The merge looked super easy so I did it in the web editor. Marking AutoMerge. If build succeeds, we're good to go.

**Comment by j4james** [comment]
> The merge looked super easy so I did it in the web editor. Marking AutoMerge. If build succeeds, we're good to go.

I think the web editor screws up line endings, which is something the code formatter now checks, and I suspect that is why it's failing.

**Comment by j4james** [comment]
@miniksa Is it OK if I just force-push a rebase? I don't know how else to fix things at this point.

**Comment by DHowett-MSFT** [comment]
@j4james that'll be fine. Sorry about that :smile:

**Comment by miniksa** [comment]
> @j4james that'll be fine. Sorry about that 😄

Yes, it's fine. Sorry, everyone stood in my office for like 45 minutes BSing so I didn't see this as it came in. 

Also, apologies that I keep learning bad merging lessons on your branches.

**Comment by j4james** [comment]
FYI, the current build failure is just the unit tests timing out, which seems to happen quite often in the CI build. I don't think there's actually anything wrong with the code itself.

**Comment by miniksa** [comment]
> FYI, the current build failure is just the unit tests timing out, which seems to happen quite often in the CI build. I don't think there's actually anything wrong with the code itself.

Yeah , looking at this in #4384

**Comment by miniksa** [comment]
WELP.

**Comment by DHowett-MSFT** [comment]
:tada: Once again, thanks for the contribution!

This pull request was included in a set of conhost changes that was just
released with [Windows Insider Build 19603](https://blogs.windows.com/windowsexperience/2020/04/08/announcing-windows-10-insider-preview-build-19603/).

**Pull request merged**